jans-pycloudlib/jans/pycloudlib/persistence/sql.py

```
@@ -669,6 +663,77 @@ def use_simple_json(self):
         sql_overrides = load_sql_overrides()
         return as_boolean(sql_overrides.get("MYSQL_SIMPLE_JSON", "true"))
 
+    def upsert(self, table_name: str, column_mapping: dict[str, _t.Any]) -> None:
+        table = self.metadata.tables.get(table_name)
+
+        # doc_id might be required for inserting new entry
+        if "doc_id" not in column_mapping:
+            raise ValueError(f"doc_id is required in column_mapping for upsert operation on {table_name}")
```

Ruff (0.14.5)

Avoid specifying long messages outside the exception class

(TRY003)